File: src/week5/P004.js
const readline = require('readline');

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});

let lines = [];

rl.on('line', (line) => {
  lines.push(line);
});

rl.on('close', () => {
  if (lines.length === 0) return;

  const n = parseInt(lines[0]);
  const graph = Array.from({ length: n + 1 }, () => []);

  for (let i = 1; i < n; i++) {
    const [a, b] = lines[i].split(' ').map(Number);
    graph[a].push(b);
    graph[b].push(a);
  }

  const visited = new Array(n + 1).fill(0);

  // DFS 함수 (재귀)
  // JavaScript는 기본 재귀 깊이 제한이 있으므로,
  // 노드 수가 많으면 스택 오버플로우가 발생할 수 있습니다.
  // 필요 시 반복문 기반 DFS로 변경해야 할 수도 있습니다.
  function dfs(s) {
    for (const next of graph[s]) {
      if (visited[next] === 0) {
        visited[next] = s;
        dfs(next);
      }
    }
  }

  // 루트 노드 1부터 시작해서 부모 노드를 찾기 위해
  // visited[1]을 1로 설정하여 다시 방문하지 않도록 함
  visited[1] = 1;
  dfs(1);

  let result = '';
  for (let x = 2; x <= n; x++) {
    result += visited[x] + '\n';
  }
  console.log(result);
});

File: src/week5/P004.java
package week5;

import java.io.*;
import java.util.*;

public class P004 {
    static ArrayList<Integer>[] graph;
    static int[] visited;

    public static void main(String[] args) throws IOException {
        // 입력 속도를 높이기 위해 BufferedReader 사용
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;

        int n = Integer.parseInt(br.readLine());

        // 그래프 초기화
        graph = new ArrayList[n + 1];
        for (int i = 1; i <= n; i++) {
            graph[i] = new ArrayList<>();
        }

        // 간선 정보 입력
        for (int i = 0; i < n - 1; i++) {
            st = new StringTokenizer(br.readLine());
            int a = Integer.parseInt(st.nextToken());
            int b = Integer.parseInt(st.nextToken());
            graph[a].add(b);
            graph[b].add(a);
        }

        visited = new int[n + 1];

        // 루트 노드 방문 처리 후 DFS 시작
        // 1번 노드의 부모는 없거나 자기 자신으로 표시하여 중복 방문 방지
        visited[1] = 1;
        dfs(1);

        // 결과 출력 (StringBuilder 사용으로 속도 향상)
        StringBuilder sb = new StringBuilder();
        for (int x = 2; x <= n; x++) {
            sb.append(visited[x]).append("\n");
        }
        System.out.print(sb);
    }

    // DFS 함수
    static void dfs(int s) {
        for (int next : graph[s]) {
            if (visited[next] == 0) {
                visited[next] = s; // 방문하지 않은 노드의 부모를 현재 노드(s)로 설정
                dfs(next);
            }
        }
    }
}
